### PR TITLE
FirebaseML - adding Long Running Operation support

### PR DESCRIPTION
**Pull request by ifielker**
Add long running operation support

### CODE REVIEW

File: src/machine-learning/machine-learning.ts [hunk 1]
```diff
@@ -246,7 +248,11 @@ export class Model {
 
   public readonly tfliteModel?: TFLiteModel;
 
-  constructor(model: ModelResponse) {
+  private readonly activeOperations?: OperationResponse[];
+  private readonly app?: FirebaseApp;
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:252]: This is a bad idea. `FirebaseApp` is a complex and heavy weight object, and attaching that onto things that are returned to the user is not going to work out well. I'd be more comfortable, if this just kept a reference to the API client instance instead. I have reservations about this approach too, but I don't see any other way around it.

Also now that this class has private fields and functions, let's also define a `toJSON()` method in case somebody tries to serialize a model object into JSON.

File: src/machine-learning/machine-learning.ts [hunk 2]
```diff
@@ -279,18 +286,55 @@ export class Model {
...
+      const client = new MachineLearningApiClient(this.app!);
+      return client.handleOperation(this.activeOperations![0], true, maxTimeSeconds)
+        .then((modelResponse) => this.updateFromResponse(modelResponse));
+    }
     return Promise.resolve();
   }
+
+  private updateFromResponse(model: ModelResponse): void {
+    // Update the readonly fields with the latest information.
+    (this.modelId as Model['modelId']) = extractModelId(model.name);
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:316]: Never a good idea to mutate readonly fields. Just store a `ModelResponse` as an attribute an expose the individual fields via getters. This method can hopefully just update the `ModelResponse` value.

```
private model: ModelResponse;

get modelId(): string {
  return this.model.modelId;
}
```

File: package.json [hunk 3]
```diff
@@ -62,7 +62,7 @@
   },
   "optionalDependencies": {
     "@google-cloud/firestore": "^4.0.0",
-    "@google-cloud/storage": "^5.0.0"
+    "@google-cloud/storage": "^5.1.2"
```
[the remark this answers is not in the document]
Comment (ifielker) [reply package.json:65]: I don't think it was, I just installed the latest, I guess it changed this file too. I'll revert.

File: src/machine-learning/machine-learning-api-client.ts [hunk 4]
```diff
@@ -80,6 +86,7 @@ export interface ListModelsResponse {
 
 export interface OperationResponse {
   readonly name?: string;
+  readonly metadata?: any;
```
[the remark this answers is not in the document]
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts]: done

File: src/machine-learning/machine-learning-api-client.ts [hunk 5]
```diff
@@ -141,14 +148,20 @@ export class MachineLearningApiClient {
...
 
-
   public getModel(modelId: string): Promise<ModelResponse> {
     return Promise.resolve()
       .then(() => {
         return this.getModelName(modelId);
       })
       .then((modelName) => {
-        return this.getResource<ModelResponse>(modelName);
+        return this.getShortNameResource<ModelResponse>(modelName);
```
[the remark this answers is not in the document]
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts]: done

File: src/machine-learning/machine-learning-api-client.ts [hunk 6]
```diff
@@ -141,14 +148,20 @@ export class MachineLearningApiClient {
...
+      });
+  }
+
+  public getOperation(operationName: string): Promise<OperationResponse> {
+    return Promise.resolve()
+      .then(() => {
+        return this.getFullNameResource<OperationResponse>(operationName);
```
[the remark this answers is not in the document]
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts]: done

File: src/machine-learning/machine-learning-api-client.ts [hunk 7]
```diff
@@ -200,15 +213,106 @@ export class MachineLearningApiClient {
...
+   * Handles a Long Running Operation coming back from the server.
+   *
+   * @param op The operation to handle
+   * @param wait Should we allow polling for the operation to complete.
+   *             If no polling is requested, a locked model will be returned instead.
+   * @param maxTimeSeconds The total maximum number of seconds to wait. 0 for default.
+   * @param baseWaitMillis The number of milliseconds to wait for the first request. Only used for testing.
+   * @param maxWaitMillis The maximum number of milliseconds to wait between requests. Only used for testing.
+   */
+  public handleOperation(
```
[the remark this answers is not in the document]
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts:222]: done

File: src/machine-learning/machine-learning-api-client.ts [hunk 8]
```diff
@@ -200,15 +213,106 @@ export class MachineLearningApiClient {
...
+
+    if (!wait) {
+      const modelId = extractModelId(metadata.name);
+      return this.getModel(modelId);
+    }
+
+    return this.pollOperationWithExponentialBackoff(opName, maxTimeSeconds, baseWaitMillis, maxWaitMillis);
+  }
+
+  private pollOperationWithExponentialBackoff(
```
[the remark this answers is not in the document]
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts:263]: done

File: src/machine-learning/machine-learning-api-client.ts [hunk 9]
```diff
@@ -200,15 +213,106 @@ export class MachineLearningApiClient {
...
+    opName: string,
+    maxTimeSeconds: number,
```
[the remark this answers is not in the document]
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts]: done

File: src/machine-learning/machine-learning-api-client.ts [hunk 10]
```diff
@@ -200,15 +213,106 @@ export class MachineLearningApiClient {
...
+    baseWaitMillis: number,
+    maxWaitMillis: number): Promise<ModelResponse> {
+    let maxTimeMilliseconds = maxTimeSeconds * 1000;
+    if (maxTimeMilliseconds <= 0) {
+      maxTimeMilliseconds = POLL_DEFAULT_MAX_TIME_MILLISECONDS;
+    }
+
+    // These should only ever be modified by unit tests to run faster.
+    if (baseWaitMillis <= 0) {
```
[the remark this answers is not in the document]
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts]: done.

File: src/machine-learning/machine-learning-api-client.ts [hunk 11]
```diff
@@ -200,15 +213,106 @@ export class MachineLearningApiClient {
...
+    return poller.poll(() => {
+      return this.getOperation(opName)
+        .then((responseData: any) => {
+          if(!responseData.done) {
+            return null;
+          }
+          if(responseData.error) {
+            const err = FirebaseMachineLearningError.fromOperationError(
+              responseData.error.code, responseData.error.message);
+            return Promise.reject(err);
```
[the remark this answers is not in the document]
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts]: done

File: src/machine-learning/machine-learning-api-client.ts [hunk 12]
```diff
@@ -200,15 +213,106 @@ export class MachineLearningApiClient {
...
+    }
+
+    const poller = new ExponentialBackoffPoller<ModelResponse>(
+      baseWaitMillis,
+      maxWaitMillis,
+      maxTimeMilliseconds);
+
+    return poller.poll(() => {
+      return this.getOperation(opName)
+        .then((responseData: any) => {
```
[the remark this answers is not in the document]
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts]: done

File: src/machine-learning/machine-learning-api-client.ts [hunk 13]
```diff
@@ -200,15 +213,106 @@ export class MachineLearningApiClient {
...
+          if(!responseData.done) {
```
[the remark this answers is not in the document]
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts]: done. (Why didn't the linter pick this up?)

File: src/machine-learning/machine-learning-api-client.ts [hunk 14]
```diff
@@ -200,15 +213,106 @@ export class MachineLearningApiClient {
...
+            const err = FirebaseMachineLearningError.fromOperationError(
+              responseData.error.code, responseData.error.message);
+            return Promise.reject(err);
+          }
+          return responseData.response;
+        });
+    });
+  }
+
+  private handleDoneOperation(op: OperationResponse): Promise<ModelResponse> {
```
[the remark this answers is not in the document]
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts]: Moved inline. It ended up only getting called once.

File: src/machine-learning/machine-learning.ts (hunk 1 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts]: Done. (Sorry, I thought client was bigger than app)

File: src/machine-learning/machine-learning.ts (hunk 2 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts]: done

File: src/machine-learning/machine-learning-api-client.ts [hunk 15]
```diff
@@ -200,15 +213,95 @@ export class MachineLearningApiClient {
...
+      throw new FirebaseMachineLearningError('invalid-server-response',
+        `Unknown Metadata type: ${JSON.stringify(metadata)}`);
+    }
+
+    if (!options || !options.wait) {
+      const modelId = extractModelId(metadata.name);
+      return this.getModel(modelId);
+    }
+
+    return this.pollOperationWithExponentialBackoff(opName, options);
```
Comment (hiranya911) [inline src/machine-learning/machine-learning-api-client.ts:259]: opName is first used here. May be just remove the variable and use `op.name!` directly?

File: src/machine-learning/machine-learning-api-client.ts [hunk 16]
```diff
@@ -200,15 +213,95 @@ export class MachineLearningApiClient {
...
+    // Operation is not done
+    const opName = op.name!;
+    const metadata = op.metadata || {};
+    const metadataType: string = metadata['@type'] || '';
+    if (!metadataType.includes('ModelOperationMetadata')) {
+      throw new FirebaseMachineLearningError('invalid-server-response',
+        `Unknown Metadata type: ${JSON.stringify(metadata)}`);
+    }
+
+    if (!options || !options.wait) {
```
Comment (hiranya911) [inline src/machine-learning/machine-learning-api-client.ts:254]:
```
if (options?.wait) {
  return this.pollOperationWithExponentialBackoff(opName, options);
}

const modelId = extractModelId(metadata.name);
return this.getModel(modelId);
```

File: src/machine-learning/machine-learning.ts [hunk 17]
```diff
@@ -229,68 +232,153 @@ export class MachineLearning implements FirebaseServiceInterface {
...
-      this.tfliteModel = {
-        gcsTfliteUri: model.tfliteModel.gcsTfliteUri,
-        sizeBytes: model.tfliteModel.sizeBytes,
-      };
+
+    if (this.tfliteModel) {
+      jsonModel['tfliteModel'] = this.tfliteModel;
     }
+
+    return JSON.stringify(jsonModel);
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:332]: I believe `toJSON()` is expected to return an object. So just return `jsonModel` here.

File: src/machine-learning/machine-learning.ts [hunk 18]
```diff
@@ -229,68 +232,153 @@ export class MachineLearning implements FirebaseServiceInterface {
...
+    if (this.model.tfliteModel) {
+      return {
+        gcsTfliteUri: this.model.tfliteModel.gcsTfliteUri,
+        sizeBytes: this.model.tfliteModel.sizeBytes,
+      };
+    }
+    return undefined;
+  }
+
+  public equals(obj: Model): boolean {
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:299]: This is not a thing in Node.js afaik. Why was this needed?

File: src/machine-learning/machine-learning.ts [hunk 19]
```diff
@@ -229,68 +232,153 @@ export class MachineLearning implements FirebaseServiceInterface {
...
-        'invalid-server-response',
-        `Invalid Model response: ${JSON.stringify(model)}`);
+  private model: ModelResponse;
+  private readonly client?: MachineLearningApiClient;
+
+  constructor(model: ModelResponse, client: MachineLearningApiClient) {
+    Model.validateModel(model);
+    this.model = model;
+    // Remove '@type' field. We don't need it.
+    if ((this.model as {[key: string]: any})["@type"]) {
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:246]: Why is this necessary? Doesn't look like it's exposed anywhere.

File: src/machine-learning/machine-learning.ts [hunk 20]
```diff
@@ -229,68 +232,153 @@ export class MachineLearning implements FirebaseServiceInterface {
...
+      // The client will always be defined on Models that have activeOperations
+      // because models with active operations came back from the server and
+      // were constructed with a non-empty client.
+      return this.client!.handleOperation(this.model.activeOperations![0], { wait: true, maxTimeMillis })
+        .then((modelResponse) => this.updateFromResponse(modelResponse));
+    }
     return Promise.resolve();
   }
+
+  private static validateModel(model: ModelResponse): void {
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:361]: validateModelResponse

File: src/machine-learning/machine-learning.ts [hunk 21]
```diff
@@ -229,68 +232,153 @@ export class MachineLearning implements FirebaseServiceInterface {
...
+
+    if (this.validationError) {
+      jsonModel['validationError'] = this.validationError;
     }
-    this.published = model.state?.published || false;
-    this.etag = model.etag;
-    if (model.modelHash) {
-      this.modelHash = model.modelHash;
+
+    if(this.modelHash) {
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:324]: Space

File: test/unit/machine-learning/machine-learning-api-client.spec.ts [hunk 22]
```diff
@@ -64,9 +64,14 @@ describe('MachineLearningApiClient', () => {
     },
   };
 
+  const PROJECT_ID = 'test-project';
+  const PROJECT_NUMBER = '1234567';
+  const OPERATION_ID = '987654';
+  const OPERATION_NAME = `projects/${PROJECT_NUMBER}/operations/${OPERATION_ID}`;
```
Comment (hiranya911) [inline test/unit/machine-learning/machine-learning-api-client.spec.ts:70]: Are we actually using project numbers in our implementation?

File: test/unit/machine-learning/machine-learning-api-client.spec.ts [hunk 23]
```diff
@@ -386,6 +415,153 @@ describe('MachineLearningApiClient', () => {
...
+      const stub = sinon
+        .stub(HttpClient.prototype, 'send')
+        .rejects(utils.errorFrom(ERROR_RESPONSE, 404));
+      stubs.push(stub);
+      const expected = new FirebaseMachineLearningError('not-found', 'Requested entity not found');
+      return apiClient.getOperation(OPERATION_NAME)
+        .should.eventually.be.rejected.and.deep.include(expected);
+    });
+
+    it('should reject unknown-error when error code is not present', () => {
```
Comment (hiranya911) [inline test/unit/machine-learning/machine-learning-api-client.spec.ts:445]: should reject with

File: test/unit/machine-learning/machine-learning-api-client.spec.ts [hunk 24]
```diff
@@ -386,6 +415,153 @@ describe('MachineLearningApiClient', () => {
...
+      const stub = sinon
+        .stub(HttpClient.prototype, 'send')
+        .rejects(utils.errorFrom({}, 404));
+      stubs.push(stub);
+      const expected = new FirebaseMachineLearningError('unknown-error', 'Unknown server error: {}');
+      return apiClient.getOperation(OPERATION_NAME)
+        .should.eventually.be.rejected.and.deep.include(expected);
+    });
+
+    it('should reject unknown-error for non-json response', () => {
```
Comment (hiranya911) [inline test/unit/machine-learning/machine-learning-api-client.spec.ts:455]: should reject with

File: src/machine-learning/machine-learning.ts [hunk 25]
```diff
@@ -229,68 +232,153 @@ export class MachineLearning implements FirebaseServiceInterface {
...
-    // This will likely change in future.
-    return false;
+    return (this.model.activeOperations?.length || 0) > 0;
   }
 
-  /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
-  public waitForUnlocked(maxTimeSeconds?: number): Promise<void> {
+  /**
+   * Wait for the active operations on the model to complete.
+   * @param maxTimeMillis The number of milliseconds to wait for the model to be unlocked. 0 for default.
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:346]: 0 for default doesn't sound right. Did you mean if unspecified a default will be used?

File: test/unit/machine-learning/machine-learning-api-client.spec.ts [hunk 26]
```diff
@@ -386,6 +415,153 @@ describe('MachineLearningApiClient', () => {
...
+    });
+
+    it('handles a running operation with wait ending in timeout', () => {
+      const stub = sinon.stub(HttpClient.prototype, 'send');
+      stub.onCall(0).resolves(utils.responseFrom(OPERATION_NOT_DONE_RESPONSE));
+      stub.onCall(1).resolves(utils.responseFrom(OPERATION_NOT_DONE_RESPONSE));
+      stub.onCall(2).resolves(utils.responseFrom(OPERATION_NOT_DONE_RESPONSE));
+      stubs.push(stub);
+      const expected = new Error('ExponentialBackoffPoller dealine exceeded - Master timeout reached');
+      return apiClient.handleOperation(OPERATION_NOT_DONE_RESPONSE, {
```
Comment (hiranya911) [inline test/unit/machine-learning/machine-learning-api-client.spec.ts:555]: Is this timing sensitive by any chance? I assume it waits 500ms before making the first call. Then makes a second call after 1s. At which point the maxTimeMillis has already passed so we get a timeout. Is that correct?

File: test/unit/machine-learning/machine-learning.spec.ts [hunk 27]
```diff
@@ -155,18 +160,79 @@ describe('MachineLearning', () => {
...
+      },
+    };
 
 
   let machineLearning: MachineLearning;
   let mockApp: FirebaseApp;
+  let mockClient: MachineLearningApiClient;
   let mockCredentialApp: FirebaseApp;
 
+  let MODEL1: Model;
```
Comment (hiranya911) [inline test/unit/machine-learning/machine-learning.spec.ts:224]: Lower-case for variables

File: test/unit/machine-learning/machine-learning.spec.ts [hunk 28]
```diff
@@ -245,6 +311,53 @@ describe('MachineLearning', () => {
...
+        updateTime: UPDATE_TIME_UTC,
+        published: true,
+        etag: 'etag123',
+        modelHash: 'modelHash123',
+        tfliteModel: {
+          gcsTfliteUri: 'gs://test-project-bucket/Firebase/ML/Models/model1.tflite',
+          sizeBytes: 16900988,
+        }
+      }
+      const expectedString = JSON.stringify(expectedModel);
```
Comment (hiranya911) [inline test/unit/machine-learning/machine-learning.spec.ts:331]: This should look something like this:

```
const jsonString = JSON.stringify(model);
expect(JSON.parse(jsonString)).to.deep.equal(expectedModel);
```

File: src/machine-learning/machine-learning.ts [hunk 29]
```diff
@@ -229,68 +232,153 @@ export class MachineLearning implements FirebaseServiceInterface {
...
-      };
+
+    if (this.tfliteModel) {
+      jsonModel['tfliteModel'] = this.tfliteModel;
     }
+
+    return JSON.stringify(jsonModel);
   }
 
+
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:335]: Move up closer to other getters.

File: src/machine-learning/machine-learning.ts [hunk 30]
```diff
@@ -229,68 +232,153 @@ export class MachineLearning implements FirebaseServiceInterface {
...
+  /**
+   * Locked indicates if there are active long running operations on the model.
+   * Models may not be modified when they are locked.
+   */
   public get locked(): boolean {
-    // Backend does not currently return locked models.
-    // This will likely change in future.
-    return false;
+    return (this.model.activeOperations?.length || 0) > 0;
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:341]: `?? 0` is safer than `|| 0` in most circumstances.

File: src/machine-learning/machine-learning-api-client.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/machine-learning/machine-learning-api-client.ts]: We might be lacking https://eslint.org/docs/rules/space-after-keywords in our rules. Will look into it separately.

**Comment by hiranya911** [comment]
Also since there's a new API here, shouldn't there be some changes in the d.ts file?

File: src/machine-learning/machine-learning.ts [hunk 31]
```diff
@@ -229,68 +232,153 @@ export class MachineLearning implements FirebaseServiceInterface {
...
+    return this.model.etag;
+  }
+
+  get modelHash(): string | undefined {
+    return this.model.modelHash;
+  }
+
+  get tfliteModel(): TFLiteModel | undefined {
+    if (this.model.tfliteModel) {
+      return {
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:291]: If there's a reason why we can't do `return this.model.tfliteModel`, let's add a comment here to indicate why.

File: src/machine-learning/machine-learning-api-client.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/machine-learning/machine-learning-api-client.ts]: Related #957

File: src/machine-learning/machine-learning-api-client.ts (hunk 15 above)
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts:259]: done

File: src/machine-learning/machine-learning-api-client.ts (hunk 16 above)
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts:254]: done

File: src/machine-learning/machine-learning.ts (hunk 17 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts:332]: done

File: src/machine-learning/machine-learning.ts (hunk 19 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts:246]: Discussed. Moved to one location instead of two.

File: src/machine-learning/machine-learning.ts (hunk 18 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts:299]: I thought it would help with the test equality, but it didn't and I forgot to remove it. Oops. (Thanks)

File: src/machine-learning/machine-learning.ts (hunk 20 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts:361]: done

File: src/machine-learning/machine-learning.ts (hunk 21 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts:324]: done

File: test/unit/machine-learning/machine-learning-api-client.spec.ts (hunk 22 above)
Comment (ifielker) [reply test/unit/machine-learning/machine-learning-api-client.spec.ts:70]: Yes, the backend returns operations with project number.

File: test/unit/machine-learning/machine-learning-api-client.spec.ts (hunk 23 above)
Comment (ifielker) [reply test/unit/machine-learning/machine-learning-api-client.spec.ts:445]: done

File: test/unit/machine-learning/machine-learning-api-client.spec.ts (hunk 24 above)
Comment (ifielker) [reply test/unit/machine-learning/machine-learning-api-client.spec.ts:455]: done

File: src/machine-learning/machine-learning.ts (hunk 25 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts:346]: Yes, missed the documentation when I updated the code.

File: test/unit/machine-learning/machine-learning-api-client.spec.ts (hunk 26 above)
Comment (ifielker) [reply test/unit/machine-learning/machine-learning-api-client.spec.ts:555]: Yes, that's correct. It should not be timing sensitive.

File: test/unit/machine-learning/machine-learning.spec.ts (hunk 27 above)
Comment (ifielker) [reply test/unit/machine-learning/machine-learning.spec.ts:224]: done

File: test/unit/machine-learning/machine-learning.spec.ts (hunk 28 above)
Comment (ifielker) [reply test/unit/machine-learning/machine-learning.spec.ts:331]: done

File: src/machine-learning/machine-learning.ts (hunk 29 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts:335]: done

File: src/machine-learning/machine-learning.ts (hunk 30 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts:341]: done

File: src/machine-learning/machine-learning.ts (hunk 31 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts:291]: Done.

File: src/machine-learning/machine-learning.ts [hunk 32]
```diff
@@ -287,24 +282,16 @@ export class Model {
   }
 
   get tfliteModel(): TFLiteModel | undefined {
+    // Make a copy so people can't directly modify the private this.model object.
     if (this.model.tfliteModel) {
-      return {
-        gcsTfliteUri: this.model.tfliteModel.gcsTfliteUri,
-        sizeBytes: this.model.tfliteModel.sizeBytes,
-      };
+      return deepCopy(this.model.tfliteModel);
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:287]: You can drop the if condition altogether. 

`deepCopy()` returns `undefined` if the input if `undefined`.

File: src/machine-learning/machine-learning.ts [hunk 33]
```diff
@@ -240,13 +240,8 @@ export class Model {
...
   constructor(model: ModelResponse, client: MachineLearningApiClient) {
-    Model.validateModel(model);
-    this.model = model;
-    // Remove '@type' field. We don't need it.
-    if ((this.model as {[key: string]: any})["@type"]) {
-      delete (this.model as {[key: string]: any})["@type"];
-    }
-
+    Model.validateModelResponse(model);
+    this.model = Model.cleanModelResponse(model);
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:244]: May be just:

```
// Combine validation and cloning/cleaning into one helper.
this.model = Model.validateAndClone(model);
```

File: src/machine-learning/machine-learning.ts [hunk 34]
```diff
@@ -229,24 +232,120 @@ export class MachineLearning implements FirebaseServiceInterface {
...
+    }
+
+    return jsonModel;
+  }
+
+  /**
+   * Locked indicates if there are active long running operations on the model.
+   * Models may not be modified when they are locked.
+   */
+  public get locked(): boolean {
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:326]: Move this up closer to other getters.

File: src/machine-learning/machine-learning.ts (hunk 32 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts:287]: done

File: src/index.d.ts [hunk 35]
```diff
@@ -1317,12 +1317,17 @@ declare namespace admin.machineLearning {
     /**
      * Wait for the model to be unlocked.
      *
-     * @param {number} maxTimeSeconds The maximum time in seconds to wait.
+     * @param {number} maxTimeMillis The maximum time in milliseconds to wait.
```
[the remark this answers is not in the document]
Comment (ifielker) [reply src/index.d.ts:1320]: done

File: src/machine-learning/machine-learning.ts (hunk 33 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts:244]: done

File: src/machine-learning/machine-learning.ts (hunk 34 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts:326]: done

File: src/machine-learning/machine-learning.ts [hunk 36]
```diff
@@ -229,68 +232,140 @@ export class MachineLearning implements FirebaseServiceInterface {
...
+   */
+  public waitForUnlocked(maxTimeMillis?: number): Promise<void> {
     // Backend does not currently return locked models.
     // This will likely change in future.
+    if ((this.model.activeOperations?.length ?? 0) > 0) {
+      // The client will always be defined on Models that have activeOperations
+      // because models with active operations came back from the server and
+      // were constructed with a non-empty client.
+      return this.client!.handleOperation(this.model.activeOperations![0], { wait: true, maxTimeMillis })
+        .then((modelResponse) => this.updateFromResponse(modelResponse));
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts:341]: Nit: updateFromResponse is now a one-liner. Just `this.model = Model.validateAndClone(modelResponse);` and remove the `updateFromResponse()` helper.

File: src/index.d.ts [hunk 37]
```diff
@@ -1317,12 +1317,18 @@ declare namespace admin.machineLearning {
...
+     *     If not specified, a default maximum of 2 minutes will be used.
```
Comment (egilmorez) [inline src/index.d.ts:1321]: We try to avoid "will."  Suggest just "is used."

File: src/index.d.ts [hunk 38]
```diff
@@ -1317,12 +1317,18 @@ declare namespace admin.machineLearning {
...
      *
      * @return {Promise<void>} A promise that resolves when the model is unlocked
      *   or the maximum wait time has passed.
      */
-    waitForUnlocked(maxTimeSeconds?: number): Promise<void>;
+    waitForUnlocked(maxTimeMillis?: number): Promise<void>;
+
+    /**
+     * Return the JSON format of the model.
```
Comment (egilmorez) [inline src/index.d.ts:1329]: This feels slightly weird, unless "format" is a special noun in your context.

Would "Return the model in JSON format" or "as a JSON object" work?

File: src/index.d.ts (hunk 37 above)
Comment (ifielker) [reply src/index.d.ts:1321]: done

File: src/index.d.ts (hunk 38 above)
Comment (ifielker) [reply src/index.d.ts:1329]: done
